[Analytics] Created Workers Analytics Engine section. (#4743)
cloudflare/cloudflare-docs: content/analytics/analytics-engine/get-started.md

---
title: Get started
pcx-content-type: how-to
weight: 1
meta:
  title: Get started with Workers Analytics Engine
---

# Get started with Workers Analytics Engine

There are three steps to get started with Workers Analytics Engine:

## 1. Configure your dataset and binding in Wrangler

All data in Workers Analytics Engine is written to a dataset. A dataset is conceptually like a table in SQL: the rows and columns should have consistent meaning.

> To access your dataset from the Workers runtime, you need to create a binding. A binding is an [environment variable](/workers/platform/environment-variables/) that you can use in the Workers runtime which enables you to write to a dataset. A dataset is created implicitly after you define your binding and begin writing to it from a Worker.

Reviewer: ```suggestion
To access your dataset from the Workers runtime, you need to create a binding. A binding is like an [environment variable](/workers/platform/environment-variables/) that you can use in the Workers runtime which enables you to write to a dataset. A dataset is created implicitly after you define your binding and begin writing to it from a Worker.
```
Environment variables are one example of a binding, but not all bindings are environment variables.

> To access your dataset from the Workers runtime, you need to create a binding. A binding is an [environment variable](/workers/platform/environment-variables/) that you can use in the Workers runtime which enables you to write to a dataset.  A dataset is created implicitly after you define your binding and begin writing to it from a Worker.

Reviewer 2: ```suggestion
To access your dataset from the Workers runtime, you need to create a binding. A binding is an [environment variable](/workers/platform/environment-variables/) that you can use in the Workers runtime which enables you to write to a dataset. A dataset is created implicitly after you define your binding and begin writing to it from a Worker.
```

A metric is just a number type field that can be aggregated in some way – for example, it could be summed, averaged, or quantiled. A label is a string type field that can be used for grouping or filtering.

For example, suppose you are collecting air quality samples. Each data point would represent a reading from your weather sensor. Metrics might include numbers like the temperature or air pressure reading. The labels could include the location of the sensor and the hardware identifier of the sensor.

This is how it translates into code:

```js
  async fetch(request: Request, env) {
    env.WEATHER.writeDataPoint({
      labels: ["Seattle", "USA", "pro_sensor_9000"],
      metrics: [25, 0.5]
    });
    return new Response("OK!");
  }
```

In our initial version, developers are responsible for **providing fields in a consistent order**, so that they have the same semantics when querying. In a future iteration, we plan to let developers name their labels and metrics in the binding, and then use these names when writing data points in the runtime.

> ## 3. Query data using GraphQL and SQL API.

Reviewer: ```suggestion
## 3. Query data using GraphQL and SQL API
```
This was the only of the three numbered headings with terminal punctuation.

Data can be queried using either [GraphQL](/analytics/graphql-api/) or SQL API.

> The GraphQL API powers our dashboard and is better suited for building interactive dashboards. At this time, the GraphQL API exposes a highly simplified schema, though we plan to support a richer schema over time.

Reviewer 2: I can submit a commit to this branch with an example added to the bottom of this file.

SQL API is better suited for writing ad hoc queries and integrating with external tools like Grafana. At this time, the SQL API only supports the `SELECT` statement and a limited subset of SQL functionality.

> The SQL API is available as an HTTP rest endpoint at `/v4/$accountTag/analytics_engine/sql` using the `POST` and `GET` method.

Reviewer: ```suggestion
The SQL API is available as an HTTP endpoint at `/v4/$accountTag/analytics_engine/sql` using the `POST` and `GET` method.
```

Or capitalize REST, if we're really using a REST architecture.

Separately, an example of how we expect SQL to be encoded in a GET method (e.g., in the URL's query parameters?) would be useful.

### Example of querying data with the SQL API

In the following example, we use the SQL API to query the top 10 cities that had the highest average humidity readings when the temperature was above zero:

```sql
SELECT label_1 as city, avg(metric_2) as avg_humidity
FROM analytics_engine
WHERE 
  dataset = 'WEATHER'
  AND metric_1 > 0
ORDER BY avg_humidity DESC
LIMIT 10
```

Note that, for our initial version, labels and metrics are accessed via names that have 1-based indexing. In the future, when we let developers name labels and metrics in their binding, these names will also be available via the SQL API.

### Working with time series

Workers Analytics Engine is optimized for powering time series analytics that can be visualized using tools like Grafana. Every event written from the runtime is automatically populated with a `timestamp` field. It is expected that most time series will round, and then `GROUP BY` the `timestamp`. For example:

```sql
SELECT

>   intDiv(datetime, 300) * 300 as t, 

Reviewer 3: ```suggestion
  intDiv(timestamp, 300) * 300 as t, 
```

  label_1 as city, 

>   avg(metric_2) as avg_humidity, 

Reviewer 3: Trailing comma:
```suggestion
  avg(metric_2) as avg_humidity
```

FROM analytics_engine
WHERE 
  dataset = 'WEATHER'
  AND timestamp >= now() - INTERVAL '1' DAY
  AND metric_1 > 0
GROUP BY t, city
ORDER BY t, avg_humidity desc
```

> This query first rounds the datetime field to the nearest five minutes. Then we group by that field and city, and calculate the average humidity in each city for a five minute period. 

Reviewer: ```suggestion
This query first rounds the `timestamp` field to the nearest five minutes. Then we group by that field and city, and calculate the average humidity in each city for a five minute period. 
```

name = "<BINDING_NAME>"
```

By default, the dataset name is the same as the binding name. If you want, you can also specify the dataset name:

```toml
[[unsafe.bindings]]
type = "analytics_engine"
name = "<BINDING_NAME>"
dataset = "<DATASET_NAME>"
```

## 2. Write data from the Workers Runtime API

Once a binding is declared in Wrangler and your worker is deployed, you get a new environment variable in the Workers runtime that represents your Workers Analytics Engine dataset. This variable has a method, `writeDataPoint()`. A data point is a structured event which consists of a vector of labels and a vector of metrics.

A metric is just a number type field that can be aggregated in some way – for example, it could be summed, averaged, or quantiled. A label is a string type field that can be used for grouping or filtering.

For example, suppose you are collecting air quality samples. Each data point would represent a reading from your weather sensor. Metrics might include numbers like the temperature or air pressure reading. The labels could include the location of the sensor and the hardware identifier of the sensor.

This is how it translates into code:

```js

>   async fetch(request: Request, env) {

Reviewer 4: ```
 [WARNING] Processing wrangler.toml configuration:

    - "unsafe" fields are experimental and may change or break at any time.

✘ [ERROR] Expected ")" but found ":"

    src/index.js:12:21:
      12 │   async fetch(request: Request, env) {
         │                      ^
         ╵                      )

✘ [ERROR] Build failed with 1 error:

  src/index.js:12:21: ERROR: Expected ")" but found ":"
  ```
  
  I get the above error when running "wrangler publish" with this code snippet in a JS worker.

Reviewer 4: ```suggestion
  async fetch(request, env) {
```

It works if I remove the `:Request`

Reviewer 5: the `:Request` is TypeScript - should be removed since this is a JS example.

  }
```

In our initial version, developers are responsible for **providing fields in a consistent order**, so that they have the same semantics when querying. In a future iteration, we plan to let developers name their labels and metrics in the binding, and then use these names when writing data points in the runtime.

## 3. Query data using GraphQL and SQL API

Data can be queried using either [GraphQL](/analytics/graphql-api/) or SQL API.

The GraphQL API powers our dashboard and is better suited for building interactive dashboards. At this time, the GraphQL API exposes a highly simplified schema, though we plan to support a richer schema over time.

SQL API is better suited for writing ad hoc queries and integrating with external tools like Grafana. At this time, the SQL API only supports the `SELECT` statement and a limited subset of SQL functionality.

The SQL API is available as an HTTP endpoint at `/v4/$accountTag/analytics_engine/sql` using the `POST` and `GET` method.

### Example of querying data with the SQL API

In the following example, we use the SQL API to query the top 10 cities that had the highest average humidity readings when the temperature was above zero:

> ```sql

Reviewer 4: ```suggestion
```sql
SELECT blob1 as city, avg(double2) as avg_humidity FROM analytics_engine WHERE dataset = 'WEATHER' AND double1 > 0 GROUP BY city ORDER BY avg_humidity DESC LIMIT 10
```

There are a number of issues with the query:
* There is a missing `GROUP BY city`
* For some reason, the query only executes if the whole statement is on a single line (I think this is a bug that needs to be fixed @46bit)

SELECT label_1 as city, avg(metric_2) as avg_humidity
FROM analytics_engine
WHERE 
  dataset = 'WEATHER'
  AND metric_1 > 0
ORDER BY avg_humidity DESC
LIMIT 10
```

Note that, for our initial version, labels and metrics are accessed via names that have 1-based indexing. In the future, when we let developers name labels and metrics in their binding, these names will also be available via the SQL API.

### Working with time series

Workers Analytics Engine is optimized for powering time series analytics that can be visualized using tools like Grafana. Every event written from the runtime is automatically populated with a `timestamp` field. It is expected that most time series will round, and then `GROUP BY` the `timestamp`. For example:

```sql
SELECT
  intDiv(toUInt32(timestamp), 300) * 300 as t, 

>   label_1 as city, 

Reviewer 4: ```suggestion
  blob1 as city, 
```

>   avg(metric_2) as avg_humidity

Reviewer 4: ```suggestion
  avg(double2) as avg_humidity
```

FROM analytics_engine
WHERE 
  dataset = 'WEATHER'
  AND timestamp >= now() - INTERVAL '1' DAY

>   AND metric_1 > 0

Reviewer 4: ```suggestion
  AND double1 > 0
```

There are three steps to get started with Workers Analytics Engine:

## 1. Configure your dataset and binding in Wrangler

All data in Workers Analytics Engine is written to a dataset. A dataset is conceptually like a table in SQL: the rows and columns should have consistent meaning.

To access your dataset from the Workers runtime, you need to create a binding using [Wrangler](https://developers.cloudflare.com/workers/wrangler/configuration/). A binding is like an [environment variable](/workers/platform/environment-variables/) that you can use in the Workers runtime which enables you to write to a dataset. A dataset is created implicitly after you define your binding and begin writing to it from a Worker.

In this guide, we will show you how to start using a dataset.

To define a binding. For example:

```toml
[[unsafe.bindings]]
type = "analytics_engine"
name = "<BINDING_NAME>"
```

By default, the dataset name is the same as the binding name. If you want, you can also specify the dataset name:

```toml
[[unsafe.bindings]]
type = "analytics_engine"
name = "<BINDING_NAME>"
dataset = "<DATASET_NAME>"
```

## 2. Write data from the Workers Runtime API

> Once a binding is declared in Wrangler and your worker is deployed, you get a new environment variable in the Workers runtime that represents your Workers Analytics Engine dataset. This variable has a method, `writeDataPoint()`. A data point is a structured event which consists of a vector of labels and a vector of metrics.

Reviewer 4: ```suggestion
Once a binding is declared in Wrangler and your worker is deployed, you get a new environment variable in the Workers runtime that represents your Workers Analytics Engine dataset. This variable has a method, `writeDataPoint()`. A data point is a structured event which consists of a vector of blobs and a vector of doubles.
```

> A metric is just a number type field that can be aggregated in some way – for example, it could be summed, averaged, or quantiled. A label is a string type field that can be used for grouping or filtering.

Reviewer 6: update

> A metric is just a number type field that can be aggregated in some way – for example, it could be summed, averaged, or quantiled. A label is a string type field that can be used for grouping or filtering.

Reviewer 4: ```suggestion
A double is just a number type field that can be aggregated in some way – for example, it could be summed, averaged, or quantiled. A blob is a string type field that can be used for grouping or filtering.
```

> For example, suppose you are collecting air quality samples. Each data point would represent a reading from your weather sensor. Metrics might include numbers like the temperature or air pressure reading. The labels could include the location of the sensor and the hardware identifier of the sensor.

Reviewer 4: ```suggestion
For example, suppose you are collecting air quality samples. Each data point would represent a reading from your weather sensor. Doubles might include numbers like the temperature or air pressure reading. The blobs could include the location of the sensor and the hardware identifier of the sensor.
```

This is how it translates into code:

```js
  async fetch(request, env) {
    env.WEATHER.writeDataPoint({

>       labels: ["Seattle", "USA", "pro_sensor_9000"],

Reviewer 6: update

>       metrics: [25, 0.5]

Reviewer 4: ```suggestion
      doubles: [25, 0.5]
```

    });
    return new Response("OK!");
  }
```

In our initial version, developers are responsible for **providing fields in a consistent order**, so that they have the same semantics when querying. In a future iteration, we plan to let developers name their labels and metrics in the binding, and then use these names when writing data points in the runtime.

## 3. Query data using GraphQL and SQL API

Data can be queried using either [GraphQL](/analytics/graphql-api/) or SQL API.

The GraphQL API powers our dashboard and is better suited for building interactive dashboards. At this time, the GraphQL API exposes a highly simplified schema, though we plan to support a richer schema over time.

SQL API is better suited for writing ad hoc queries and integrating with external tools like Grafana. At this time, the SQL API only supports the `SELECT` statement and a limited subset of SQL functionality.

> The SQL API is available as an HTTP endpoint at `/v4/$accountTag/analytics_engine/sql` using the `POST` and `GET` method.

Reviewer 3: ```suggestion
The SQL API is available as an HTTP endpoint at `/v4/$accountTag/analytics_engine/sql` using the `POST` and `GET` method. You need to supply a Cloudflare [API Token](https://dash.cloudflare.com/profile/api-tokens) that has the `Account Analytics Read` permission.
```

### Example of querying data with the SQL API

> In the following example, we use the SQL API to query the top 10 cities that had the highest average humidity readings when the temperature was above zero:

Reviewer 3: ```suggestion
In the following example, we use the SQL API to query the top 10 cities that had the highest average humidity readings when the temperature was above zero.
```

```sql

> SELECT blob1 as city, avg(double2) as avg_humidity FROM analytics_engine WHERE dataset = 'WEATHER' AND double1 > 0 GROUP BY city ORDER BY avg_humidity DESC LIMIT 10

Reviewer 6: update

> SELECT label_1 as city, avg(metric_2) as avg_humidity

Reviewer 6: This SQL query should be deleted

FROM analytics_engine
WHERE 
  dataset = 'WEATHER'
  AND metric_1 > 0
ORDER BY avg_humidity DESC
LIMIT 10
```

> Note that, for our initial version, labels and metrics are accessed via names that have 1-based indexing. In the future, when we let developers name labels and metrics in their binding, these names will also be available via the SQL API.

Reviewer 6: update labels/metrics

### Working with time series

Workers Analytics Engine is optimized for powering time series analytics that can be visualized using tools like Grafana. Every event written from the runtime is automatically populated with a `timestamp` field. It is expected that most time series will round, and then `GROUP BY` the `timestamp`. For example:

> ```sql

Reviewer 4: The code block below also needs to be a single line too

SELECT
  intDiv(toUInt32(timestamp), 300) * 300 as t, 

>   blob1 as city, 

Reviewer 6: update

```toml
[[unsafe.bindings]]
type = "analytics_engine"
name = "<BINDING_NAME>"
dataset = "<DATASET_NAME>"
```

## 2. Write data from the Workers Runtime API

Once a binding is declared in Wrangler and your worker is deployed, you get a new environment variable in the Workers runtime that represents your Workers Analytics Engine dataset. This variable has a method, `writeDataPoint()`. A data point is a structured event which consists of a vector of blobs and a vector of doubles.

A double is just a number type field that can be aggregated in some way – for example, it could be summed, averaged, or quantiled. A blob is a string type field that can be used for grouping or filtering.

For example, suppose you are collecting air quality samples. Each data point would represent a reading from your weather sensor. Doubles might include numbers like the temperature or air pressure reading. The blobs could include the location of the sensor and the hardware identifier of the sensor.

This is how it translates into code:

```js
  async fetch(request, env) {
    env.WEATHER.writeDataPoint({

>       blobs: ["Seattle", "USA", "pro_sensor_9000"],

Reviewer 6: ```suggestion
      'blobs': ["Seattle", "USA", "pro_sensor_9000"],
```

>       doubles: [25, 0.5]

Reviewer 6: ```suggestion
      'doubles': [25, 0.5]
```

    });
    return new Response("OK!");
  }
```

> In our initial version, developers are responsible for **providing fields in a consistent order**, so that they have the same semantics when querying. In a future iteration, we plan to let developers name their labels and metrics in the binding, and then use these names when writing data points in the runtime.

Reviewer 6: ```suggestion
In our initial version, developers are responsible for **providing fields in a consistent order**, so that they have the same semantics when querying. In a future iteration, we plan to let developers name their blobs and doubles in the binding, and then use these names when writing data points in the runtime.
```

The GraphQL API powers our dashboard and is better suited for building interactive dashboards. At this time, the GraphQL API exposes a highly simplified schema, though we plan to support a richer schema over time.

SQL API is better suited for writing ad hoc queries and integrating with external tools like Grafana. At this time, the SQL API only supports the `SELECT` statement and a limited subset of SQL functionality.

The SQL API is available as an HTTP endpoint at `/v4/$accountTag/analytics_engine/sql` using the `POST` and `GET` method. You need to include an `Authorization: Bearer _____` token where the underscores should be replaced with a Cloudflare [API Token](https://dash.cloudflare.com/profile/api-tokens) that has the `Account Analytics Read` permission.

### Example of querying data with the SQL API

In the following example, we use the SQL API to query the top 10 cities that had the highest average humidity readings when the temperature was above zero.

Here's how we represent that as SQL:

```sql
SELECT blob1 as city, avg(double2) as avg_humidity FROM analytics_engine WHERE dataset = 'WEATHER' AND double1 > 0 GROUP BY city ORDER BY avg_humidity DESC LIMIT 10

> SELECT label_1 as city, avg(metric_2) as avg_humidity

Reviewer 6: ```suggestion
```

> FROM analytics_engine

Reviewer 6: ```suggestion
```

> WHERE 

Reviewer 6: ```suggestion
```

>   dataset = 'WEATHER'

Reviewer 6: ```suggestion
```

>   AND metric_1 > 0

Reviewer 6: ```suggestion
```

> ORDER BY avg_humidity DESC

Reviewer 6: ```suggestion
```

> LIMIT 10

Reviewer 6: ```suggestion
```

> LIMIT 10

Reviewer 6: ```suggestion
```

> ```

Reviewer 6: ```suggestion
```

> Note that, for our initial version, labels and metrics are accessed via names that have 1-based indexing. In the future, when we let developers name labels and metrics in their binding, these names will also be available via the SQL API.

Reviewer 6: ```suggestion
Note that, for our initial version, blobs and doubles are accessed via names that have 1-based indexing. In the future, when we let developers name blobs and doubles in their binding, these names will also be available via the SQL API.
```

---
title: Get started
pcx-content-type: how-to
weight: 1
meta:
  title: Get started with Workers Analytics Engine
---

# Get started with Workers Analytics Engine

There are three steps to get started with Workers Analytics Engine:

## 1. Configure your dataset and binding in Wrangler

> All data in Workers Analytics Engine is written to a dataset. A dataset is conceptually like a “table” in SQL; the rows and columns should have consistent meaning.

Reviewer 7: ```suggestion
All data in Workers Analytics Engine is written to a dataset. A dataset is conceptually like a table in SQL: the rows and columns should have consistent meaning.
```

Reviewer 7: I think in one of our content critiques, we decided on not using quotes in this way

To get started using the Workers Analytics Engine, you need to create a dataset and a binding that associates a given Worker with a dataset. This is done by editing [wrangler.toml](/workers/wrangler/configuration/). For example:

```toml
[analytics_engine]
bindings = [
    { name = "<EVENT_NAME>" }
]

> ```

Reviewer 3: We'll need to implement that in the Wrangler CLI. As of right now the syntax is:

```toml
[[unsafe.bindings]]
type = "analytics_engine"
name = "<BINDING_NAME>"
```

That defaults to the dataset name being the same as the binding name, but you can also specify the dataset name:

```
[[unsafe.bindings]]
type = "analytics_engine"
name = "<BINDING_NAME>"
dataset = "<DATASET_NAME>"
```

Not sure whether to actually change the docs, we should PR wrangler instead?

## 2. Write data from the Workers Runtime API

Once a binding is declared in Wrangler and your worker is deployed, you get a new [environment variable](/workers/platform/environment-variables/) in the Workers runtime that represents your Workers Analytics Engine dataset. This variable has a method, `writeDataPoint()`. A data point is a structured event which consists of a vector of labels and a vector of metrics.

A metric is just a number type field that can be aggregated in some way – for example, it could be summed, averaged, or quantiled. A label is a string type field that can be used for grouping or filtering.

For example, suppose you are collecting air quality samples. Each data point would represent a reading from your weather sensor. Metrics might include numbers like the temperature or air pressure reading. The labels could include the location of the sensor and the hardware identifier of the sensor.

This is how it translates into code:

> ```bash

Reviewer 5: ```suggestion
```js
```

  async fetch(request: Request, env) {
    env.WEATHER.writeDataPoint({

>       labels: ["Seattle", "USA", "pro_sensor_9000”],

Reviewer 8: Fixed good old mac smart quotes
```suggestion
      labels: ["Seattle", "USA", "pro_sensor_9000"],
```

      metrics: [25, 0.5]
    });
    return new Response("OK!");
  }
```

In our initial version, developers are responsible for **providing fields in a consistent order**, so that they have the same semantics when querying. In a future iteration, we plan to let developers name their labels and metrics in the binding, and then use these names when writing data points in the runtime.

## 3. Query data using GraphQL and SQL API.

> Data can be queried using either GraphQL or SQL API.

Reviewer 5: ```suggestion
Data can be queried using either [GraphQL](/analytics/graphql-api/) or SQL API.
```

> The GraphQL UI powers our dashboard and is better suited for building interactive dashboards. At this time, the GraphQL API exposes a highly simplified schema, though we plan to support a richer schema over time.

Reviewer 5: ```suggestion
The GraphQL API powers our dashboard and is better suited for building interactive dashboards. At this time, the GraphQL API exposes a highly simplified schema, though we plan to support a richer schema over time.
```

SQL API is better suited for writing ad hoc queries and integrating with external tools like Grafana. At this time, the SQL API only supports the `SELECT` statement and a limited subset of SQL functionality.

The SQL API is available as an HTTP rest endpoint at `/v4/$accountTag/analytics_engine/sql` using the `POST` and `GET` method.

### Example of querying data with the SQL API

> In the following example, we use the SQL API to query the top 10 cities that had the highest average humidity readings when the temperature was above 0:

Reviewer 7: ```suggestion
In the following example, we use the SQL API to query the top 10 cities that had the highest average humidity readings when the temperature was above zero:
```

```bash

> SELECT label_1 as city, avg(metric_2) as avg_humidity

Reviewer 5: @46bit are these column names correct? I don't recall there being an underscore.

FROM analytics_engine
WHERE 
  dataset = 'WEATHER'
  AND metric_1 > 0
ORDER BY avg_humidity DESC
LIMIT 10
```

Note that, for our initial version, labels and metrics are accessed via names that have 1-based indexing. In the future, when we let developers name labels and metrics in their binding, these names will also be available via the SQL API.

### Working with time series

> Workers Analytics Engine is optimized for powering time series analytics that can be visualized using tools like Grafana. Every event written from the runtime is automatically populated with a `datetime` field. It is expected that most time series will round, and then GROUP BY the `datetime`. For example:

Reviewer 5: ```suggestion
Workers Analytics Engine is optimized for powering time series analytics that can be visualized using tools like Grafana. Every event written from the runtime is automatically populated with a `timestamp` field. It is expected that most time series will round, and then GROUP BY the `timestamp`. For example:
```

> ```bash

Reviewer 5: ```suggestion
```sql
```

> Workers Analytics Engine is optimized for powering time series analytics that can be visualized using tools like Grafana. Every event written from the runtime is automatically populated with a `timestamp` field. It is expected that most time series will round, and then GROUP BY the `timestamp`. For example:

Reviewer 7: I think `GROUP BY` is an SQL statement, and should be in between backticks. If not, please change to small caps as this makes it seem like we're shouting at the reader :)

SELECT
  intDiv(datetime, 300) * 300 as t, 
  label_1 as city, 
  avg(metric_2) as avg_humidity, 
FROM analytics_engine
WHERE 
  dataset = 'WEATHER'
  AND timestamp >= now() - 86400
  AND metric_1 > 0
GROUP BY t, city
ORDER BY t, avg_humidity desc
```

> This query first rounds the datetime field to the nearest five minutes. Then we group by that field and city and calculate the average humidity in each city for a five minute period. 

Reviewer 7: ```suggestion
This query first rounds the datetime field to the nearest five minutes. Then we group by that field and city, and calculate the average humidity in each city for a five minute period. 
```
